components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,192 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/learning_service.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/logging.h"
+#include "base/task/thread_pool.h"
+#include "base/time/time.h"
+#include "base/timer/timer.h"
+#include "brave/components/brave_federated/communication_adapter.h"
+#include "brave/components/brave_federated/eligibility_service.h"
+#include "brave/components/brave_federated/features.h"
+#include "brave/components/brave_federated/task/federated_task_runner.h"
+#include "brave/components/brave_federated/task/model.h"
+#include "brave/components/brave_federated/task/typing.h"
+#include "brave/components/brave_federated/util/synthetic_dataset.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "net/base/backoff_entry.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+
+namespace brave_federated {
+
+LearningService::LearningService(
+    EligibilityService* eligibility_service,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : url_loader_factory_(url_loader_factory),
+      eligibility_service_(eligibility_service) {
+  DCHECK(!init_task_timer_);
+  const int init_federated_service_wait_time_in_seconds =
+      brave_federated::features::GetInitFederatedServiceWaitTimeInSeconds();
+
+  post_results_policy_ = std::make_unique<const net::BackoffEntry::Policy>(
+      /*.num_errors_to_ignore = */ 0,
+      /*.initial_delay_ms = */
+      features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.multiply_factor =*/2.0,
+      /*.jitter_factor =*/0.0,
+      /*.maximum_backoff_ms =*/16 *
+          features::GetFederatedLearningUpdateCycleInSeconds() * 1000,
+      /*.always_use_initial_delay =*/true);
+  post_results_backoff_entry_ =
+      std::make_unique<net::BackoffEntry>(post_results_policy_.get());
+
+  init_task_timer_ = std::make_unique<base::OneShotTimer>();
+  init_task_timer_->Start(
+      FROM_HERE, base::Seconds(init_federated_service_wait_time_in_seconds),
+      this, &LearningService::Init);
+}
+
+LearningService::~LearningService() {
+  StopParticipating();
+  eligibility_service_->RemoveObserver(this);
+}
+
+void LearningService::Init() {
+  DCHECK(url_loader_factory_);
+  DCHECK(eligibility_service_);
+  DCHECK(init_task_timer_);
+  communication_adapter_ =
+      std::make_unique<CommunicationAdapter>(url_loader_factory_);
+  eligibility_service_->AddObserver(this);
+
+  StartParticipating();
+}
+
+void LearningService::StartParticipating() {
```

Comment:
Is it possible to have logically overlapped calls of this function? I mean another `StartParticipating` call then the previous pipeline isn't finished.

Reply:
no, it shouldn't be possible. Unless the eligibility service is jittery (a state change is notified when there wasn't one), for that I added a check on the `participating_` flag

Reply:
Let's consider a situation:
1. `StartParticipating()` call, GetTasks() and the next pipeline is in progress.
2. `StopParticipating()` call
3. another `StartParticipating()` call, but the first processing isn't finished.

Reply:
oh, okay. In that sense yes. They will overlap. Would the best solution be to finish with the current pipeline and not start a new one, or interrupt the initial pipeline at whatever stage on `StopParticipating` and start a new one? The latter might be more cumbersome to implement. Wdyt?